marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java

```diff
@@ -114,4 +116,52 @@ private String convertRowToJSONString(InternalRow record) {
         return jsonObjectWriter.toString();
     }
 
+    protected JsonNode determineIngestionEndpointParams(Map<String, String> params) {
+        ObjectMapper objectMapper = new ObjectMapper();
+
+        String ingestEndpointParams = params.containsKey("ingestendpointparams")?
+            params.get("ingestendpointparams"):objectMapper.createObjectNode().toString();
+
+        JsonNode endpointParamsJsonNode;
+        try {
+            endpointParamsJsonNode = objectMapper.readValue(ingestEndpointParams, JsonNode.class);
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to parse ingestendpointparams, cause: " + e.getMessage(), e);
+        }
+
+        boolean doesNotHaveApiPath = !endpointParamsJsonNode.has("apiPath");
+        boolean hasWorkUnitOrEndpointState = endpointParamsJsonNode.has("workUnit") || endpointParamsJsonNode.has("endpointState");
+        if (doesNotHaveApiPath && hasWorkUnitOrEndpointState) {
+            throw new RuntimeException("Cannot set workUnit or endpointState in ingestionendpointparams unless apiPath is defined as well.");
+        }
+
+        final String apiModulePath = (endpointParamsJsonNode.hasNonNull("apiPath") &&
+            endpointParamsJsonNode.get("apiPath").asText().length() > 0) ?
+            endpointParamsJsonNode.get("apiPath").asText() : "/data-hub/5/data-services/ingestion/bulkIngester.api";
+        logger.info("Will write to endpoint defined by: " + apiModulePath);
+        ((ObjectNode)endpointParamsJsonNode).put("apiPath", apiModulePath);
+
+        if (endpointParamsJsonNode.hasNonNull("endpointState") && endpointParamsJsonNode.get("endpointState").asText().length() > 0) {
+            ((ObjectNode)endpointParamsJsonNode).set("endpointState", endpointParamsJsonNode.get("endpointState"));
+        }
+        // TODO : remove the below else block after java-client-api 5.3 release
+        else {
+            ((ObjectNode)endpointParamsJsonNode).put("endpointState", "{}");
+        }
+
+        if (endpointParamsJsonNode.hasNonNull("workUnit") && endpointParamsJsonNode.get("workUnit").asText().length() > 0) {
+
+            JsonNode workUnitNode;
+            try {
+                workUnitNode = objectMapper.readValue(endpointParamsJsonNode.get("workUnit").asText(), JsonNode.class);
+            } catch (JsonProcessingException e) {
+                throw new RuntimeException("Unable to parse workUnit, cause: " + e.getMessage(), e);
+            }
+            ((ObjectNode) workUnitNode).put("uriprefix", params.get("uriprefix"));
+            ((ObjectNode)endpointParamsJsonNode).set("workUnit", workUnitNode);
+        } else {
+            ((ObjectNode)endpointParamsJsonNode).put("workUnit", "{\"uriprefix\":\"" + params.get("uriprefix") + "\"}");
```

Comment:
I think this needs to be a JsonNode as well instead of a string for consistency. You can do mapper.createObjectNode() and then node.put("uriprefix", params.get("uriprefix")) . 

I think it's good to handle these as JsonNode's instead of as strings. When Ernie provides us with a custom workUnit and/or endpointState, we should convert those ASAP to JsonNode. That way, if Ernie's JSON is malformed, it'll fail as quickly as possible - i.e. in the connector code and not in ML. 

Also this is where @SameeraPriyathamTadikonda will add his code for collections/permissions/etc.